handler.go

```diff
@@ -0,0 +1,318 @@
+	streamTransport, ok := t.(transport.StreamTransport)
+	if !ok {
+		return fmt.Errorf("Transport does not support stream calls: %q", t.Protocol())
+	}
+
+	ctx, cancel := tchannel.NewContext(streamReq.Request.Timeout)
+	defer cancel()
+	ctx = makeContextWithTrace(ctx, t, streamReq.Request, 0)
+
+	stream, err := streamTransport.CallStream(ctx, streamReq)
+	if err != nil {
+		return fmt.Errorf("Failed while making stream call: %v", err)
+	}
+
+	rpcType := serializer.MethodType()
+	if rpcType == encoding.BidirectionalStream {
+		return makeBidiStream(ctx, stream, reqSupplier, resHandler)
+	} else if rpcType == encoding.ClientStream {
+		return makeClientStream(ctx, stream, reqSupplier, resHandler)
+	}
+	return makeServerStream(ctx, stream, reqSupplier, resHandler)
+}
+
+// makeServerStream starts server-side streaming rpc
+func makeServerStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	req, err := reqSupplier()
+	if err != nil && err != io.EOF {
+		return err
+	}
+	if err == nil {
+		// verify there is no second request
+		_, err = reqSupplier()
+		if err == nil {
+			return fmt.Errorf("Request data contains more than 1 message for server-streaming RPC")
+		} else if err != io.EOF {
+			return err
+		}
+	}
+
+	err = sendStreamMessage(ctx, stream, req)
+	if err != nil {
+		return err
+	}
+
+	for err == nil {
+		var resBody []byte
+		resBody, err = receiveStreamMessage(ctx, stream)
+		if err != nil {
+			break
+		}
+		err = resHandler(resBody)
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+// makeClientStream starts client-side streaming rpc
+func makeClientStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	var err error
+	for err == nil {
+		var reqBody []byte
+		reqBody, err = reqSupplier()
+		if err != nil {
+			break
+		}
+		err = sendStreamMessage(ctx, stream, reqBody)
+	}
+
+	if err == io.EOF {
+		err = closeSendStream(ctx, stream)
+	}
+	if err != nil {
+		return err
+	}
+
+	res, err := receiveStreamMessage(ctx, stream)
+	if err != nil {
+		return err
+	}
+	return resHandler(res)
+}
+
+// makeBidiStream starts bi-directional streaming rpc
+func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	var wg sync.WaitGroup
+	var sendErr error
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg.Add(1)
+	// start go routine to concurrently send stream messages
+	go func() {
+		defer wg.Done()
+
+		var err error
+		for err == nil {
+			var reqBody []byte
+			reqBody, err = reqSupplier()
+			if err == io.EOF {
+				err = closeSendStream(ctx, stream)
+				break
+			}
+			if err != nil {
+				cancel()
+				break
+			}
+
+			err = sendStreamMessage(ctx, stream, reqBody)
+		}
```

Comment:
Is there any chance to reuse the code from `makeClientStream` ? Also is there any reason those code are not the same since they have the same behavior?

You should also add a TODO that the message sent will be controlled by a timer once the option --stream-interval is available

Reply:
Yes, both are different behaviour to input handling, io.EOF on response, etc.

| Client stream      | Server stream |
| ----------- | ----------- |
| 0 or more requests      | mandatory one request       |
| -  | error if more than one request body in input        |
| mandatory one response | 0 or more response |
| error if receive response is EOF, since one response is mandatory |  - |

Reply:
Yes the confusion was about the 0 request you explained in another comment